subsys/ipc/ipc_service/backends/ipc_icmsg.c

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <device.h>
+#include <string.h>
+#include <drivers/mbox.h>
+
+#include <ipc/ipc_service_backend.h>
+#include "ipc_icmsg.h"
+
+#define DT_DRV_COMPAT	zephyr_ipc_icmsg
+#define CB_BUF_SIZE	CONFIG_IPC_SERVICE_BACKEND_ICMSG_CB_BUF_SIZE
+
+static const uint8_t magic[] = {0x45, 0x6d, 0x31, 0x6c, 0x31, 0x4b,
+				0x30, 0x72, 0x6e, 0x33, 0x6c, 0x69, 0x34};
+BUILD_ASSERT(sizeof(magic) <= CB_BUF_SIZE);
+BUILD_ASSERT(CB_BUF_SIZE <= UINT8_MAX);
+
+struct backend_data_t {
+	/* Tx/Rx buffers. */
+	struct icmsg_buf *tx_ib;
+	struct icmsg_buf *rx_ib;
+
+	/* Backend ops for an endpoint. */
+	const struct ipc_service_cb *ops;
+
+	/* General */
+	struct k_work mbox_work;
+	atomic_t state;
+};
+
+struct backend_config_t {
+	uintptr_t tx_shm_addr;
+	uintptr_t rx_shm_addr;
+	size_t tx_shm_size;
+	size_t rx_shm_size;
+	struct mbox_channel mbox_tx;
+	struct mbox_channel mbox_rx;
+};
+
+
+static void mbox_callback_process(struct k_work *item)
+{
+	struct backend_data_t *dev_data = CONTAINER_OF(item, struct backend_data_t, mbox_work);
+	uint8_t cb_buffer[CB_BUF_SIZE] __aligned(4);
+
+	atomic_t state = atomic_get(&dev_data->state);
+	int len = icmsg_buf_read(dev_data->rx_ib, cb_buffer, CB_BUF_SIZE);
+
+	__ASSERT_NO_MSG(len <= CB_BUF_SIZE);
+
+	if (len == -EAGAIN) {
+		(void)k_work_submit(&dev_data->mbox_work);
+		return;
+	} else if (len <= 0) {
+		return;
+	}
+
+	if (state == ICMSG_STATE_READY) {
+		if (dev_data->ops && dev_data->ops->received) {
+			dev_data->ops->received(cb_buffer, len, NULL);
+		}
+
+		/* Reading with NULL buffer to know if there are data in the buffer to be read. */
+		len = icmsg_buf_read(dev_data->rx_ib, NULL, 0);
+		if (len > 0) {
+			(void)k_work_submit(&dev_data->mbox_work);
+		}
+	} else {
+		__ASSERT_NO_MSG(state == ICMSG_STATE_BUSY);
+		if (len != sizeof(magic) || memcmp(magic, cb_buffer, len)) {
+			__ASSERT_NO_MSG(false);
+			return;
+		}
+
+		if (dev_data->ops && dev_data->ops->received) {
```

Comment:
Probably copy-paste error
```suggestion
		if (dev_data->ops && dev_data->ops->bound) {
```

Reply:
whopsie